File: App/src/main/scala/Tutorials.scala
trait Tutorials[F[_]] {
  def insert(tutorial: TutorialInput): F[String]
  def update(tutorial: Tutorial): F[String]
  def delete(id: String): F[Int]

  def getAll: F[List[Tutorial]]
  def getById(id: String): F[Option[Tutorial]]
  def deleteAll(): F[Unit]
  def findByKeyWord(keyword: String): F[List[Tutorial]]
}

object Tutorials {
  implicit def apply[F[_]](implicit ev: Tutorials[F]): Tutorials[F] = ev
  
  final class TutorialImpl[F[_]](service: TutorialService[F]) extends Tutorials[F] {
    override def insert(tutorial: TutorialInput): F[String] = service.insert(tutorial)

    override def update(tutorial: Tutorial): F[String] = service.update(tutorial)

    override def delete(id: String): F[Int] = service.delete(id)

    override def getAll: F[List[Tutorial]] = service.getAll

    override def getById(id: String): F[Option[Tutorial]] = service.getById(id)

    override def deleteAll(): F[Unit] = service.deleteAll()

    override def findByKeyWord(keyword: String): F[List[Tutorial]] = service.findByKeyword(keyword)
  }

  def impl[F[_]](svc: TutorialService[F]): Tutorials[F] = new TutorialImpl[F](svc)
}

File: frontend/src/actions/tutorials.js
import Api from '../services/tutorial.service'
import {CREATE_TUTORIAL, DELETE_ALL_TUTORIALS, DELETE_TUTORIAL, RETRIEVE_TUTORIALS, UPDATE_TUTORIAL} from "./types";

export const createTutorial = (title, description) => async (dispatch) => {
  try {
    const res = await Api.create({title, description});
    dispatch({
      type: CREATE_TUTORIAL,
      payload: res.data
    })
    return Promise.resolve(res.data)
  } catch (err) {
    return Promise.reject(err)
  }
}

export const retrieveTutorial = () => async (dispatch) => {
  try {
    const res = await Api.getAll();
    dispatch({
      type: RETRIEVE_TUTORIALS,
      payload: res.data
    })
    return Promise.resolve(res.data)
  } catch (err) {
    return Promise.reject(err)
  }
}

export const updateTutorial = (id, data) => async (dispatch) => {
  try {
    const res = await Api.update(id, data);
    dispatch({
      type: UPDATE_TUTORIAL,
      payload: res.data
    })
    return Promise.resolve(res.data)
  } catch (err) {
    return Promise.reject(err)
  }
}

export const deleteTutorial = (id) => async (dispatch) => {
  try {
    const res = await Api.delete(id);
    dispatch({
      type: DELETE_TUTORIAL,
      payload: res.data
    })
    return Promise.resolve(res.data)
  } catch (err) {
    return Promise.reject(err)
  }
}

export const deleteTutorials = () => async (dispatch) => {
  try {
    const res = await Api.deleteAll();
    dispatch({
      type: DELETE_ALL_TUTORIALS,
      payload: res.data
    })
    return Promise.resolve(res.data)
  } catch (err) {
    return Promise.reject(err)
  }
}

export const findTutorialsByKeyWord = (keyWord) => async (dispatch) => {
  try {
    const res = await Api.findByKeyWord(keyWord);
    dispatch({
      type: RETRIEVE_TUTORIALS,
      payload: res.data
    })
    return Promise.resolve(res.data)
  } catch (err) {
    return Promise.reject(err)
  }
}